isis/src/control/objs/BundleSolutionInfo/BundleSolutionInfo.cpp

```diff
@@ -1161,82 +1085,113 @@ namespace Isis {
     QList<Statistics> rmsImageResiduals = m_statisticsResults->rmsImageResiduals();
 
-    QString ofname = "bundleout_images.csv";
-    ofname = m_settings->outputFilePrefix() + ofname;
-    m_csvSavedImagesFilename = ofname;
+    bool errorProp = false;
+    if (m_statisticsResults->converged() && m_settings->errorPropagation()) {
+      errorProp = true;
+    }
 
-    std::ofstream fpOut(ofname.toLatin1().data(), std::ios::out);
-    if (!fpOut) {
-      return false;
+    QList<QString> outputCsvFileNames;
+    QList<QString> instrumentIds = m_statisticsResults->observations().instrumentIds();
+    // If there's just a single instrumentId just call it bundleout_images.csv
+    if (instrumentIds.size() == 1) {
+      QString ofname = "bundleout_images.csv";
+      ofname = m_settings->outputFilePrefix() + ofname;
+      m_csvSavedImagesFilename = ofname;
+      outputCsvFileNames.push_back(ofname);
+    }
+    // Otherwise append the instrument IDs so it's bundleout_images_spacecraft_sensor.csv
+    else {
+      for (int i = 0; i < instrumentIds.size(); i++) {
+        QString updatedInstrumentId = instrumentIds[i];
+        // Replace and "/" or " " characters with "_" to make the filename safer
+        // This line must be separate to avoid modifying the instrumentId in the list
+        // we will iterate over later
+        updatedInstrumentId.replace("/", "_").replace(" ", "_");
+        QString ofname = "bundleout_images_" + updatedInstrumentId + ".csv";
+        ofname = m_settings->outputFilePrefix() + ofname;
+        m_csvSavedImagesFilename = ofname;
+        outputCsvFileNames.push_back(ofname);
+      }
     }
 
+    for (int i = 0; i < instrumentIds.size(); i++) {
 
-    AbstractBundleObservationQsp observation;
+      std::ofstream fpOut(outputCsvFileNames[i].toLatin1().data(), std::ios::out);
+      if (!fpOut) {
+        return false;
+      }
 
-    int nObservations = m_statisticsResults->observations().size();
+      QList<AbstractBundleObservationQsp> observations =
+          m_statisticsResults->observations().observationsByInstId(instrumentIds[i]);
 
-    outputImagesCSVHeader(fpOut);
+      int nObservations = observations.size();
 
-    bool errorProp = false;
-    if (m_statisticsResults->converged() && m_settings->errorPropagation()) {
-      errorProp = true;
-    }
+      outputImagesCSVHeader(fpOut, observations.front());
 
-    for (int i = 0; i < nObservations;i++ ) {
-      observation = m_statisticsResults->observations().at(i);
 
-      if(!observation) {
-        continue;
-      }
 
-      int numImages = observation->size();
+      for (int j = 0; j < nObservations; j++ ) {
+        AbstractBundleObservationQsp observation = observations[j];
 
-      for (int j = 0; j < numImages; j++) {
+        if(!observation) {
+          continue;
+        }
 
-        BundleImageQsp image = observation->at(j);
+        int numImages = observation->size();
 
+        for (int k = 0; k < numImages; k++) {
```

Comment:
This triple nested for loop isn't as bad as it looks because collectively it still just iterates over all the images. We just split the images up into 3 different tiers of sets. First we separate by instrument ID, then by observation, then by image.